vault/logical_cubbyhole.go

```diff
@@ -0,0 +1,194 @@
+package vault
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/vault/logical"
+	"github.com/hashicorp/vault/logical/framework"
+)
+
+// CubbyholeBackendFactory constructs a new cubbyhole backend
+func CubbyholeBackendFactory(conf *logical.BackendConfig) (logical.Backend, error) {
+	var b CubbyholeBackend
+	b.Backend = &framework.Backend{
+		Help: strings.TrimSpace(cubbyholeHelp),
+
+		Paths: []*framework.Path{
+			&framework.Path{
+				Pattern: ".*",
+
+				Callbacks: map[logical.Operation]framework.OperationFunc{
+					logical.ReadOperation:   b.handleRead,
+					logical.WriteOperation:  b.handleWrite,
+					logical.DeleteOperation: b.handleDelete,
+					logical.ListOperation:   b.handleList,
+				},
+
+				HelpSynopsis:    strings.TrimSpace(cubbyholeHelpSynopsis),
+				HelpDescription: strings.TrimSpace(cubbyholeHelpDescription),
+			},
+		},
+	}
+
+	if conf == nil {
+		return nil, fmt.Errorf("Configuation passed into backend is nil")
+	}
+	b.Backend.Setup(conf)
+
+	return b, nil
+}
+
+// CubbyholeBackend is used for storing secrets directly into the physical
+// backend. The secrets are encrypted in the durable storage.
+// This differs from generic in that every token has its own private
+// storage view. The view is removed when the token expires.
+type CubbyholeBackend struct {
+	*framework.Backend
+}
+
+func (b *CubbyholeBackend) revoke(saltedToken string, storageView logical.Storage) error {
+	if saltedToken == "" {
+		return fmt.Errorf("[ERR] cubbyhole: client token empty during revocation")
+	}
+	// Delete the entire tree in a stupid fashion for the moment
+	// to avoid changing the Storage interface
+	keys, err := storageView.List(saltedToken + "/")
```

Comment:
Much easier is to use the `ClearView` helper. You can do `ClearView(storageView.SubView(saltedToken+"/"))`